metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala

```diff
@@ -1,902 +1,774 @@
-    buildClient,
-    tables,
-    bspGlobalDirectories,
-    clientConfig.initialConfig,
-  )
-
-  private val bspConnector: BspConnector = new BspConnector(
-    bloopServers,
-    bspServers,
-    buildTools,
-    languageClient,
-    tables,
-    () => userConfig,
-    statusBar,
-    bspConfigGenerator,
-    () => bspSession.map(_.mainConnection),
-  )
-
-  private val workspaceSymbols: WorkspaceSymbolProvider =
-    new WorkspaceSymbolProvider(
-      workspace,
-      buildTargets,
-      definitionIndex,
-      saveClassFileToDisk = !clientConfig.isVirtualDocumentSupported(),
-      () => excludedPackageHandler,
-      classpathSearchIndexer = classpathSearchIndexer,
+  override def onTypeFormatting(
+      params: DocumentOnTypeFormattingParams
+  ): CompletableFuture[ju.List[TextEdit]] =
+    execForWorkspaceFolder(
+      params.getTextDocument.getUri(),
+      _.onTypeFormatting(params),
+      List.empty.asJava,
     )
 
-  private val definitionProvider: DefinitionProvider = new DefinitionProvider(
-    workspace,
-    mtags,
-    buffers,
-    definitionIndex,
-    semanticdbs,
-    warnings,
-    () => compilers,
-    remote,
-    trees,
-    buildTargets,
-    scalaVersionSelector,
-    saveDefFileToDisk = !clientConfig.isVirtualDocumentSupported(),
-    sourceMapper,
-    workspaceSymbols,
-  )
-
-  val stacktraceAnalyzer: StacktraceAnalyzer = new StacktraceAnalyzer(
-    workspace,
-    buffers,
-    definitionProvider,
-    clientConfig.icons,
-    clientConfig.commandInHtmlFormat(),
-  )
-
-  private val testProvider: TestSuitesProvider = new TestSuitesProvider(
-    buildTargets,
-    buildTargetClasses,
-    trees,
-    definitionIndex,
-    semanticdbs,
-    buffers,
-    clientConfig,
-    () => userConfig,
-    languageClient,
-  )
-
-  private val codeLensProvider: CodeLensProvider = {
-    val runTestLensProvider =
-      new RunTestCodeLens(
-        buildTargetClasses,
-        buffers,
-        buildTargets,
-        clientConfig,
-        () => userConfig,
-        trees,
-        workspace,
-      )
-    val goSuperLensProvider = new SuperMethodCodeLens(
-      buffers,
-      () => userConfig,
-      clientConfig,
-      trees,
+  override def rangeFormatting(
+      params: DocumentRangeFormattingParams
+  ): CompletableFuture[ju.List[TextEdit]] =
+    execForWorkspaceFolder(
+      params.getTextDocument.getUri(),
+      _.rangeFormatting(params),
+      List.empty.asJava,
     )
-    val worksheetCodeLens = new WorksheetCodeLens(clientConfig)
-
-    new CodeLensProvider(
-      codeLensProviders = List(
-        runTestLensProvider,
-        goSuperLensProvider,
-        worksheetCodeLens,
-        testProvider,
-      ),
-      semanticdbs,
-      stacktraceAnalyzer,
+
+  override def prepareRename(
+      params: TextDocumentPositionParams
+  ): CompletableFuture[lsp4j.Range] =
+    execForWorkspaceFolder(
+      params.getTextDocument.getUri(),
+      _.prepareRename(params),
+      new lsp4j.Range(
```

Comment:
We could maybe instead have an exception? `"workspace not found for file"`

Or alternatively we could have a default workspace that handles single files, but that could come later.

Reply:
So.. for now, we want to throw an exception in all cases not just here, right?

Reply:
Hmm.... actually coming to think of it, this would make single files not work at all, so maybe we should just get the first workspace folder? In that case it will be treated as a standalone file.

Reply:
How does this work right now? It feels like we should rather add an artificial workspace, otherwise things like completions from workspace or fallback "go-to" would be confusing. We can make a separate task out of this and for now take the first one.